### ISSUE TITLE
Opening zarr store sequentially in write mode overwrites store

### ISSUE THREAD

**Issue by jdevng** [zarr-developers/zarr-python#1427]
### Zarr version

2.14.2

### Numcodecs version

0.11

### Python Version

3.10.9

### Operating System

Mac OSX 

### Installation

pip into miniconda

### Description

import zarr
# Open store
z1 = zarr.open('data/example.zarr', mode='w', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
z1[0,0] =41
# Confirm value has been written
print("z1:" + str(z1[0,0]))

# Open same store
z2 = zarr.open('data/example.zarr', mode='w', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
# Value is zero
print("z2:" + str(z2[0,0]))

# Original value is also zero
print("z1:" + str(z1[0,0]))

>python zarrMulti.py
z1:41
z2:0
z1:0

>pip show zarr

Name: zarr
Version: 2.14.2
Summary: An implementation of chunked, compressed, N-dimensional arrays for Python

### Steps to reproduce

Run script provided

### Additional output

>python zarrMulti.py
z1:41
z2:0
z1:0

This shows the value of the store that is written, and then loading the zarr store reflects that the data was either not stored, or the data is being overwritten.

**Comment by MSanKeys963** [comment]
Thanks for helping out here, @d-v-b.

Did it work @jdevng?

**Comment by jdevng** [comment]
Ah, thank you! - I interpreted 'overwrite' as 'append' in terms of the file. I interpreted 'overwrite' as when the array is updated in RAM, would then update the store on the filesystem. This would be opposed to updating in memory, and then pushing a series of edits to the store on the filesystem. Looks like overwrite optionally erases the file contents at the beginning.

There are multiple entry points into overwrite and the file mode - create, open_array, group, open_group. 

The functon open_* uses mode - where the others use overwrite. So overwrite actually refers to the same context as mode - when first opening the file, and erases the existing data.

Isn't overwrite implied by the use of the mode? 
w - always erase file data
a - do not erase file data
r - do not erase file data
etc

Why is overwrite there? Why are they not defined on the same functions, but mutually exclusive?

**Comment by rabernat** [comment]
I agree that our API uses `mode` and `overwrite` in an inconsistent and confusing way. See my comment in https://github.com/zarr-developers/zarr-python/pull/1309#pullrequestreview-1231599350.

IMO we are overdue for a major refactor of all of this logic.

**Comment by d-v-b** [comment]
seconding the suggestion from @rabernat.
